### PR TITLE
Make E0121's suggestion more robust (+ fix E0308's suggestion)

### LINKED ISSUE #80844: E0121 suggests the use of the unique type of a generator
cc #80179

This still fails on generators ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7ac611ca1bd560116d4b9a804817ad20)):

```rust
#![feature(generators)]

fn foo() -> _ {
    || yield 4i32
}
```

```
error[E0121]: the type placeholder `_` is not allowed within types on item signatures
 --> src/lib.rs:3:13
  |
3 | fn foo() -> _ {
  |             ^
  |             |
  |             not allowed in type signatures
  |             help: replace with the correct return type: [generator@src/lib.rs:4:5: 4:18 {i32, ()}]
```

Tested on:
```
cargo 1.50.0-nightly (75d5d8cff 2020-12-22)
release: 1.50.0
commit-hash: 75d5d8cffe3464631f82dcd3c470b78dc1dda8bb
commit-date: 2020-12-22
```

I'm working on a PR to fix this as well as making this suggestion and E0308's `try adding a return type` suggestion use the same mechanism for checking if a type is suggestable directly.

@rustbot claim
@rustbot modify labels: +A-diagnostics +D-invalid-suggestion +F-generators +T-compiler

### PR DESCRIPTION
**Pull request by PatchMixolydic**
Before, both E0308 and E0121 suggested using the unnameable type of a generator as a return type. This fixes this by making `Ty::is_suggestable` return `false` for generators and by making E0121 use `Ty::is_suggestable` to prevent this from happening in the future.

Fixes #80844.

### CODE REVIEW

**Comment by estebank** [comment]
The only thing that jumps up to me is that `is_suggestable` should actually be a recursive check, so we don't suggest something like `fn() -> [generator]` or `Foo<[closure]>` by accident. Would you want to deal with that in this PR? Otherwise, r=me.

@bors delegate+

**Comment by PatchMixolydic** [comment]
I've made the check recursive for algebraic data types, arrays, slices, function pointers, raw pointers, references, and tuples, which seems to cover every `TyKind` with accessible types. I'll push once I have the test suite passing + blessed.

**Comment by PatchMixolydic** [comment]
!? I rebased onto master before pushing; not sure how this happened...

**Comment by PatchMixolydic** [comment]
Not sure if I should approve this yet since it seems to cause a diagnostics regression with generic ADTs:
```console
$ cargo +nightly -Vv
cargo 1.50.0-nightly (75d5d8cff 2020-12-22)
release: 1.50.0
commit-hash: 75d5d8cffe3464631f82dcd3c470b78dc1dda8bb
commit-date: 2020-12-22
$ cargo +nightly check
    Checking rust-issue v0.1.0 (/tmp/rust-issue)
error[E0308]: mismatched types
 --> src/main.rs:2:5
  |
2 |     Some(0i32)
  |     ^^^^^^^^^^ expected `()`, found enum `Option`
  |
  = note: expected unit type `()`
                  found enum `Option<i32>`
help: try adding a semicolon
  |
2 |     Some(0i32);
  |               ^
help: try adding a return type
  |
1 | fn expected_unit_got_option_i32() -> Option<i32> {
  |                                   ^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rust-issue`

To learn more, run the command again with --verbose.
$ cargo +stage1 check
    Checking rust-issue v0.1.0 (/tmp/rust-issue)
error[E0308]: mismatched types
 --> src/main.rs:2:5
  |
1 | fn expected_unit_got_option_i32() {
  |                                   - possibly return type missing here?
2 |     Some(0i32)
  |     ^^^^^^^^^^- help: try adding a semicolon: `;`
  |     |
  |     expected `()`, found enum `Option`
  |
  = note: expected unit type `()`
                  found enum `Option<i32>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rust-issue`

To learn more, run the command again with --verbose.
```

Note that the suggestion is missing with this PR applied. I'm not sure if that's a blocker, but it might cause issues for rust-analyzer or other such tools.

**Comment by estebank** [comment]
Oh boy! That's quite a tangle you got there! Git is so hard to use :-/ I've caused similar problems in the past to the point that I've dropped PRs and started again.

You might be able to get away with doing a `git checkout` from a commit before the merge in your `git reflog`.

File: compiler/rustc_typeck/src/collect.rs [hunk 1]
```diff
@@ -1545,7 +1545,7 @@ fn fn_sig(tcx: TyCtxt<'_>, def_id: DefId) -> ty::PolyFnSig<'_> {
                     let ret_ty = fn_sig.output();
                     if ret_ty != tcx.ty_error() {
```
Comment (estebank) [inline compiler/rustc_typeck/src/collect.rs:1546]: I think this check can now be removed thanks to `is_suggestable` accounting for `Error`.

**Comment by estebank** [comment]
> Note that the suggestion is missing with this PR applied. I'm not sure if that's a blocker, but it might cause issues for rust-analyzer or other such tools.

It's surprising that the test with `struct Container<T>(T);` works but this test case with `Option<T>` doesn't :-/

Can you add debug statements to see what the type parameters are? It could be that at this point it hasn't figured out that it would be `Option<i32>` and it is an `Option<T>`, which is why it doesn't suggest it. You _might_ have to call `resolve_vars_if_possible` and that could make it work, but can't tell for sure.

**Comment by PatchMixolydic** [comment]
> > Note that the suggestion is missing with this PR applied. I'm not sure if that's a blocker, but it might cause issues for rust-analyzer or other such tools.
> 
> It's surprising that the test with `struct Container<T>(T);` works but this test case with `Option<T>` doesn't :-/
> 
> Can you add debug statements to see what the type parameters are? It could be that at this point it hasn't figured out that it would be `Option<i32>` and it is an `Option<T>`, which is why it doesn't suggest it. You _might_ have to call `resolve_vars_if_possible` and that could make it work, but can't tell for sure.

I have; it seems that it hasn't resolved types yet (I tried using `bool` to make sure it wasn't just `{integer}` weirdness). I'll try seeing if `resolve_vars_if_possible` helps matters.

**Comment by PatchMixolydic** [comment]
???

**Comment by PatchMixolydic** [comment]
I'm not entirely sure how to get `InferCtxt::resolve_vars_if_possible` to map over the fields in a variant (assuming that's what you meant). It seems like I can map them over the fields' `DefId`s, but I'm not sure how to get the field's type from there.

**Comment by JohnCSimon** [comment]
Ping from triage - @PatchMixolydic
can you please resolve the merge conflict and build failure?

@Rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by crlf0710** [comment]
@PatchMixolydic Ping from triage, CI is still red here.

**Comment by PatchMixolydic** [comment]
Sorry about the inactivity; I've been fairly busy with college and general burnout. I'll get on this in a couple days, time permitting.

**Comment by PatchMixolydic** [comment]
Quick update: I'm planning on fixing the build issues tomorrow. Might also look into fixing the test failure, but I'm still concerned that it might be above my head. Once again, sorry about the absences. Life has been rough lately :sweat:

**Comment by PatchMixolydic** [comment]
Oh dear, that's quite a few more test failures than I expected :cold_sweat:

Going to go to bed (haven't slept in over 24 hours) and take a glance at this after I wake up, but I might be in over my head. It might ultimately be better for me (or someone smarter than me) to cherrypick/replicate these changes on a new branch and make a new PR for that.

**Comment by PatchMixolydic** [comment]
> take a glance at this after I wake up

sleep defects are hard

Looking at the test failures, it seems like something fundamental broke. I hope that this might just need a manual rebase, but as you might be able to guess at this point I'm probably not the best person for the job due to my extremely warped schedule ~~and sense of time~~. I'd like to leave these commits up as reference for anyone else who tries to fix it, but I'm not sure if I should leave the PR open or close it. I apologize for taking up so much time with this.

@rustbot modify labels +S-inactive

**Comment by JohnCSimon** [comment]
Ping from triage: @PatchMixolydic 
- I see this is still marked as inactive, do you want to close or reach out for help elsewhere? 
Thank you!

**Comment by JohnCSimon** [comment]
On second thought, I'll close this

**Pull request closed without merging**